File: app/src/test/java/com/coinninja/coinkeeper/view/adapter/util/BindableTransactionTest.java
package com.coinninja.coinkeeper.view.adapter.util;

import com.coinninja.coinkeeper.model.helpers.WalletHelper;
import com.coinninja.coinkeeper.util.currency.USDCurrency;
import com.coinninja.coinkeeper.view.adapter.util.BindableTransaction.SendState;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import static org.hamcrest.CoreMatchers.equalTo;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
public class BindableTransactionTest {

    @Mock
    WalletHelper walletHelper;
    private BindableTransaction bindableTransaction;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);
        when(walletHelper.getLatestPrice()).thenReturn(new USDCurrency(1000.00d));
        bindableTransaction = new BindableTransaction(walletHelper);
    }

    @Test
    public void returns_name_as_priority() {
        bindableTransaction.setContactName("--name--");
        bindableTransaction.setContactPhoneNumber("--number--");
        bindableTransaction.setTargetAddress("--address--");

        bindableTransaction.getIdentifiableTarget();
        assertThat(bindableTransaction.getIdentifiableTarget(), equalTo("--name--"));
    }

    @Test
    public void returns_number_formatted() {
        bindableTransaction.setContactName("");
        bindableTransaction.setContactPhoneNumber("+13305551111");
        bindableTransaction.setTargetAddress("--address--");

        bindableTransaction.getIdentifiableTarget();
        assertThat(bindableTransaction.getIdentifiableTarget(), equalTo("(330) 555-1111"));
    }

    @Test
    public void returns_number_when_name_not_available() {
        bindableTransaction.setContactName("");
        bindableTransaction.setContactPhoneNumber("+13305551111");
        bindableTransaction.setTargetAddress("--address--");

        bindableTransaction.getIdentifiableTarget();
        assertThat(bindableTransaction.getIdentifiableTarget(), equalTo("(330) 555-1111"));
    }

    @Test
    @Config(qualifiers = "es-rAU")
    public void returns_number_when_name_not_available_as_international_format() {
        bindableTransaction.setContactName("");
        bindableTransaction.setContactPhoneNumber("+13305551111");
        bindableTransaction.setTargetAddress("--address--");

        bindableTransaction.getIdentifiableTarget();
        assertThat(bindableTransaction.getIdentifiableTarget(), equalTo("+1 330-555-1111"));
    }

    @Test
    public void returns_address_when_no_contact_data() {
        bindableTransaction.setContactName("");
        bindableTransaction.setContactPhoneNumber("");
        bindableTransaction.setTargetAddress("--address--");

        bindableTransaction.getIdentifiableTarget();
        assertThat(bindableTransaction.getIdentifiableTarget(), equalTo("--address--"));
    }

    @Test
    public void calculatesTotalCryptoForSendState() {
        bindableTransaction.setFee(100L);
        bindableTransaction.setValue(1000L);

        bindableTransaction.setSendState(SendState.SEND);
        assertThat(bindableTransaction.totalCryptoForSendState().toLong(), equalTo(1100L));

        bindableTransaction.setSendState(SendState.RECEIVE);
        assertThat(bindableTransaction.totalCryptoForSendState().toLong(), equalTo(1000L));

        bindableTransaction.setSendState(SendState.TRANSFER);
        assertThat(bindableTransaction.totalCryptoForSendState().toLong(), equalTo(100L));
    }

    @Test
    public void calculatesTotalFiatForSendState() {
        bindableTransaction.setFee(1000L);
        bindableTransaction.setValue(100000L);

        bindableTransaction.setSendState(SendState.SEND);
        assertThat(bindableTransaction.totalFiatForSendState().toFormattedString(), equalTo("1.01"));

        bindableTransaction.setSendState(SendState.RECEIVE);
        assertThat(bindableTransaction.totalFiatForSendState().toFormattedString(), equalTo("1"));

        bindableTransaction.setSendState(SendState.TRANSFER);
        assertThat(bindableTransaction.totalFiatForSendState().toFormattedString(), equalTo("0.01"));
    }

    @Test
    public void simplifies_send_state__basic_direction() {
        bindableTransaction.setSendState(SendState.SEND);
        assertThat(bindableTransaction.getBasicDirection(), equalTo(SendState.SEND));
        bindableTransaction.setSendState(SendState.SEND_CANCELED);
        assertThat(bindableTransaction.getBasicDirection(), equalTo(SendState.SEND));
        bindableTransaction.setSendState(SendState.FAILED_TO_BROADCAST_SEND);
        assertThat(bindableTransaction.getBasicDirection(), equalTo(SendState.SEND));

        bindableTransaction.setSendState(SendState.RECEIVE);
        assertThat(bindableTransaction.getBasicDirection(), equalTo(SendState.RECEIVE));
        bindableTransaction.setSendState(SendState.RECEIVE_CANCELED);
        assertThat(bindableTransaction.getBasicDirection(), equalTo(SendState.RECEIVE));
        bindableTransaction.setSendState(SendState.FAILED_TO_BROADCAST_RECEIVE);
        assertThat(bindableTransaction.getBasicDirection(), equalTo(SendState.RECEIVE));

        bindableTransaction.setSendState(SendState.TRANSFER);
        assertThat(bindableTransaction.getBasicDirection(), equalTo(SendState.TRANSFER));
        bindableTransaction.setSendState(SendState.FAILED_TO_BROADCAST_TRANSFER);
        assertThat(bindableTransaction.getBasicDirection(), equalTo(SendState.TRANSFER));
    }
}

File: app/src/test/java/com/coinninja/coinkeeper/view/adapter/util/BindableTransactionTest.kt
package com.coinninja.coinkeeper.view.adapter.util

import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.coinninja.coinkeeper.model.helpers.WalletHelper
import com.coinninja.coinkeeper.util.currency.USDCurrency
import com.coinninja.coinkeeper.view.adapter.util.BindableTransaction.SendState
import com.nhaarman.mockitokotlin2.whenever
import org.hamcrest.CoreMatchers.equalTo
import org.hamcrest.MatcherAssert.assertThat
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mockito.mock

@RunWith(AndroidJUnit4::class)
class BindableTransactionTest {

    private fun createBindableTransaction(): BindableTransaction {
        val bindableTransaction = BindableTransaction(ApplicationProvider.getApplicationContext(), mock(WalletHelper::class.java))
        whenever(bindableTransaction.walletHelper.latestPrice).thenReturn(USDCurrency(1000.00))
        return bindableTransaction
    }

    @Test
    fun `identifiable target is send to self for transfers`() {
        val bindableTransaction = createBindableTransaction()
        bindableTransaction.targetAddress = "--address--"
        bindableTransaction.sendState = SendState.TRANSFER
        bindableTransaction.identity = "some body"

        assertThat(bindableTransaction.identifiableTarget, equalTo("Sent to myself"))

        bindableTransaction.sendState = SendState.SEND
        assertThat(bindableTransaction.identifiableTarget, equalTo("some body"))

        bindableTransaction.identity = ""
        assertThat(bindableTransaction.identifiableTarget, equalTo("--address--"))
    }

    @Test
    fun calculatesTotalCryptoForSendState() {
        val bindableTransaction = createBindableTransaction()
        bindableTransaction.fee = 100L
        bindableTransaction.value = 1000L

        bindableTransaction.sendState = SendState.SEND
        assertThat(bindableTransaction.totalCryptoForSendState().toLong(), equalTo(1100L))

        bindableTransaction.sendState = SendState.RECEIVE
        assertThat(bindableTransaction.totalCryptoForSendState().toLong(), equalTo(1000L))

        bindableTransaction.sendState = SendState.TRANSFER
        assertThat(bindableTransaction.totalCryptoForSendState().toLong(), equalTo(100L))
    }

    @Test
    fun calculatesTotalFiatForSendState() {
        val bindableTransaction = createBindableTransaction()
        bindableTransaction.fee = 1000L
        bindableTransaction.value = 100000L

        bindableTransaction.sendState = SendState.SEND
        assertThat(bindableTransaction.totalFiatForSendState().toFormattedString(), equalTo("1.01"))

        bindableTransaction.sendState = SendState.RECEIVE
        assertThat(bindableTransaction.totalFiatForSendState().toFormattedString(), equalTo("1"))

        bindableTransaction.sendState = SendState.TRANSFER
        assertThat(bindableTransaction.totalFiatForSendState().toFormattedString(), equalTo("0.01"))
    }

    @Test
    fun simplifies_send_state__basic_direction() {
        val bindableTransaction = createBindableTransaction()
        bindableTransaction.sendState = SendState.SEND
        assertThat(bindableTransaction.basicDirection, equalTo(SendState.SEND))
        bindableTransaction.sendState = SendState.SEND_CANCELED
        assertThat(bindableTransaction.basicDirection, equalTo(SendState.SEND))
        bindableTransaction.sendState = SendState.FAILED_TO_BROADCAST_SEND
        assertThat(bindableTransaction.basicDirection, equalTo(SendState.SEND))

        bindableTransaction.sendState = SendState.RECEIVE
        assertThat(bindableTransaction.basicDirection, equalTo(SendState.RECEIVE))
        bindableTransaction.sendState = SendState.RECEIVE_CANCELED
        assertThat(bindableTransaction.basicDirection, equalTo(SendState.RECEIVE))
        bindableTransaction.sendState = SendState.FAILED_TO_BROADCAST_RECEIVE
        assertThat(bindableTransaction.basicDirection, equalTo(SendState.RECEIVE))

        bindableTransaction.sendState = SendState.TRANSFER
        assertThat(bindableTransaction.basicDirection, equalTo(SendState.TRANSFER))
        bindableTransaction.sendState = SendState.FAILED_TO_BROADCAST_TRANSFER
        assertThat(bindableTransaction.basicDirection, equalTo(SendState.TRANSFER))
    }
}